emulator/src/hdmi_cec_driver.c

```diff
@@ -0,0 +1,828 @@
+
+typedef enum {
+  HAL_STATE_CLOSED = 0,
+  HAL_STATE_OPEN,
+  HAL_STATE_READY
+} cec_hal_state_t;
+
+typedef enum {
+  CEC_EVENT_HOTPLUG = 1,
+  CEC_EVENT_COMMAND,
+  CEC_EVENT_CONFIG
+} cec_event_type_t;
+
+typedef struct {
+  bool connected;
+  unsigned int port_id;
+} hotplug_event_t;
+
+typedef enum{
+  CEC_ACTIVE_SOURCE = 0x82,
+  CEC_IMAGE_VIEW_ON = 0x04,
+  CEC_TEXT_VIEW_ON = 0x0D,
+  CEC_INACTIVE_SOURCE = 0x9D,
+  CEC_REQUEST_ACTIVE_SOURCE = 0x85,
+  CEC_STANDBY = 0x36,
+  CEC_UNKNOWN = 0
+} cec_command_t;
+
+
+typedef struct {
+  unsigned char src_logical_addr;
+  unsigned char dest_logical_addr;
+  cec_command_t command;
+} cec_command_event_t;
+
+
+typedef struct {
+  cec_event_type_t type;
+  union {
+    cec_command_event_t cec_cmd;
+    hotplug_event_t hot_plug;
+  };
+} cec_event_t;
+
+typedef struct {
+   device_type_t type;
+   cec_version_t version;
+   unsigned int physical_address;
+   unsigned int logical_address;
+   bool active_source;
+   unsigned int vendor_id;
+   device_power_status_t power_status;
+   char osd_name[MAX_OSD_NAME_LENGTH];
+} device_info_t;
+
+typedef struct {
+  HdmiCecRxCallback_t rx_cb_func;
+  void* rx_cb_data;
+  HdmiCecTxCallback_t tx_cb_func;
+  void* tx_cb_data;
+} cec_callbacks_t;
+
+typedef struct {
+  cec_hal_state_t state;
+  device_info_t emulated_device;
+  int num_ports;
+  hdmi_port_info_t *ports;
+  int num_devices;
+  device_info_t* connected_devices;
+
+  cec_callbacks_t callbacks;
+  //TODO
+  // Eventing Queue, Thread for callback
+
+} hdmi_cec_t;
+
+
+/**Emulator Data types*/
+typedef struct {
+  unsigned short cp_port;
+  char* cp_path;
+  hdmi_cec_t * cec_hal;
+} cec_emulator_t;
+
+static cec_emulator_t* gEmulator = NULL;
+
+const static struct  {
+  device_type_t val;
+  char* str_type;
+} gDIMap [] = {
+  { DEVICE_TYPE_TV, "TV" },
+  { DEVICE_TYPE_PLAYBACK, "PlaybackDevice" },
+  { DEVICE_TYPE_AUDIO_SYSTEM, "AudioSystem" },
+  { DEVICE_TYPE_RECORDER, "RecordingDevice" },
+  { DEVICE_TYPE_TUNER, "Tuner" },
+  { DEVICE_TYPE_RESERVED, "Reserved" }
+};
+
+const static struct  {
+    char* name;
+    unsigned int code;
+  } gVCMap [] = {
+    {"TOSHIBA", 0x000039},
+    {"SAMSUNG", 0x0000F0},
+    {"DENON", 0x0005CD},
+    {"MARANTZ", 0x000678},
+    {"LOEWE", 0x000982},
+    {"ONKYO", 0x0009B0},
+    {"MEDION", 0x000CB8},
+    {"TOSHIBA2", 0x000CE7},
+    {"APPLE", 0x0010FA},
+    {"HARMAN_KARDON2", 0x001950},
+    {"GOOGLE", 0x001A11},
+    {"AKAI", 0x0020C7},
+    {"AOC", 0x002467},
+    {"PANASONIC", 0x008045},
+    {"PHILIPS", 0x00903E},
+    {"DAEWOO", 0x009053},
+    {"YAMAHA", 0x00A0DE},
+    {"GRUNDIG", 0x00D0D5},
+    {"PIONEER", 0x00E036},
+    {"LG", 0x00E091},
+    {"SHARP", 0x08001F},
+    {"SONY", 0x080046},
+    {"BROADCOM", 0x18C086},
+    {"SHARP2", 0x534850},
+    {"VIZIO", 0x6B746D},
+    {"BENQ", 0x8065E9},
+    {"HARMAN_KARDON", 0x9C645E},
+    {"UNKNOWN", 0},
+  };
+
+const static struct  {
+  device_power_status_t status;
+  char* str;
+} gPSMap [] = {
```

Comment:
don't do this its horrid coding standard

the definition & the construction seperate.